### PR TITLE
Add tidb_low_resolution_tso session scope variable on master

### PR DESCRIPTION
**Pull request by sunxiaoguang**
Low resolution timestamp is updated once every two seconds which can be used if reading slightly old version of data is acceptable. Like tidb_snapshot read, transaction is readonly when tidb_low_resolution_tso is enabled.

### What problem does this PR solve? 
Introduce a session scope variable tidb_low_resolution_tso which can use periodically updated
timestamp from TSO when transaction starts. This can save one TSO request roundtrip time per transaction which is beneficial for small readonly queries in terms of latency. As a side effect, there will be less TSO requests hence less pressure on PD.

### What is changed and how it works?
Add a new tidb_low_resolution_tso session scope variable. When transaction starts and about to get start timestamp from TSO, call low resolution timestamp interface when tidb_low_resolution_tso is enabled.   Like tidb_snapshot read, transaction is readonly when tidb_low_resolution_tso is enabled.

Tests 
 - Unit test
 - Manual test (add detailed scripts or steps below)

Related changes
 - Need to update the documentation

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -376,6 +376,9 @@ type SessionVars struct {
 
 	// PessimisticLock indicates whether new transaction should be pessimistic .
 	PessimisticLock bool
+
+	// LowResolutionTSO is used for reading data with low resolution TSO which is updated once every two seconds
```
Comment (lamxTyler) [inline sessionctx/variable/session.go]: Add `.` at the end of sentences.

File: session/txn.go [hunk 2]
```diff
@@ -381,8 +381,9 @@ func mergeToDirtyDB(dirtyDB *executor.DirtyDB, op dirtyTableOperation) {
 
 // txnFuture is a promise, which promises to return a txn in future.
 type txnFuture struct {
-	future oracle.Future
-	store  kv.Storage
+	future        oracle.Future
+	store         kv.Storage
+	lowResolution bool
```
Comment (lamxTyler) [inline session/txn.go]: Where is it used?

File: store/tikv/oracle/oracles/pd.go [hunk 3]
```diff
@@ -140,3 +140,21 @@ func (o *pdOracle) updateTS(ctx context.Context, interval time.Duration) {
 func (o *pdOracle) Close() {
 	close(o.quit)
 }
+
+type lowResolutionTsFuture uint64
+
+// Wait implements the oracle.Future interface.
+func (f lowResolutionTsFuture) Wait() (uint64, error) {
```
[the remark this answers is not in the document]
Comment (sunxiaoguang) [reply store/tikv/oracle/oracles/pd.go]: It does look verbose now. But i'm not sure about Rez, looks strange to me though. Let's see what other people think about it.

File: session/txn.go (hunk 2 above)
Comment (sunxiaoguang) [reply session/txn.go]: Oops. It was used earlier to detect update at commit time. Not useful anymore. Let me remove it. Thanks.

File: sessionctx/variable/session.go (hunk 1 above)
Comment (sunxiaoguang) [reply sessionctx/variable/session.go]: Fixed

File: store/tikv/oracle/oracles/pd.go [hunk 4]
```diff
@@ -140,3 +140,21 @@ func (o *pdOracle) updateTS(ctx context.Context, interval time.Duration) {
 func (o *pdOracle) Close() {
 	close(o.quit)
 }
+
+type lowResolutionTsFuture uint64
```
[the remark this answers is not in the document]
Comment (sunxiaoguang) [reply store/tikv/oracle/oracles/pd.go]: Fixed

**Comment by sunxiaoguang** [comment]
It's strange that I have signed CLA however it kept complaining it's not done yet.

File: store/tikv/oracle/oracles/pd.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ngaut) [reply store/tikv/oracle/oracles/pd.go]: I prefer to keep the whole word.

**Comment by sunxiaoguang** [comment]
Finally got CLA passed after merging master

**Comment by tiancaiamao** [comment]
LGTM

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -426,6 +429,7 @@ func NewSessionVars() *SessionVars {
 		CommandValue:                uint32(mysql.ComSleep),
 		TiDBOptJoinReorderThreshold: DefTiDBOptJoinReorderThreshold,
 		SlowQueryFile:               config.GetGlobalConfig().Log.SlowQueryFile,
+		LowResolutionTSO:            false,
```
Comment (zz-jason) [inline sessionctx/variable/session.go]: no need to set it to false explicitly, golang init it with false automatically.

Comment (sunxiaoguang) [reply sessionctx/variable/session.go]: Fixed

**Pull request merged**